spec/lib/saml_idp/request_spec.rb

```diff
@@ -397,6 +397,60 @@ module SamlIdp
       end
     end
 
+    describe '#matching_cert' do
+      let(:saml_request) { make_saml_request }
+
+      subject do
+        described_class.from_deflated_request saml_request
+      end
+
+      describe 'document is not signed' do
+        it 'returns nil' do
+          expect(subject.matching_cert).to be nil
+        end
+      end
+
+      describe 'document is signed' do
+        let(:saml_request) { signed_auth_request }
+        let(:service_provider)  { subject.service_provider }
+        let(:cert) { saml_settings.get_sp_cert }
+
+        describe 'the service provider has no registered certs' do
+          it 'returns nil' do
+            expect(subject.matching_cert).to be nil
+          end
+        end
+
+        describe 'the service provider has one registered certs' do
+          before { subject.service_provider.certs = [cert] }
+
+          describe 'the cert matches the assertion cert' do
+            it 'returns the cert' do
+              expect(subject.matching_cert).to eq cert
+            end
+          end
+
+          describe 'the cert does not match the assertion cert' do
+            let(:cert) { OpenSSL::X509::Certificate.new(cloudhsm_idp_x509_cert) }
+            it 'returns nil' do
+              expect(subject.matching_cert).to be nil
+            end
+          end
+
+        end
+
+        describe 'multiple certs' do
+          let(:not_matching_cert) { OpenSSL::X509::Certificate.new(cloudhsm_idp_x509_cert) }
+
+          before { subject.service_provider.certs = [not_matching_cert, invalid_cert, cert] }
```

Comment:
This could just be my lack of familiarity but it is odd that the 3 certificates come from 3 different places. I would expect them all to be local to this file or come from `spec/support/certificate_helpers.rb`.

Reply:
it is pretty odd. 

the certs are used in different places all over the spec files -- i do want to clean this up more (have already done some pretty enormous amounts of refactoring in the test files) but requires more investigation, and would make this pr a bunch messier, so i'd prefer to do it in a different PR.

Reply:
as a bit of historical context, the cloudhsm one probably comes from code that has since been removed to add encrypting SAML payloads via external service, so clearly there's still room to clean up more things left from that